src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,240 @@
+            /// The target of this reference. This value is initialized to a non-<see langword="null"/> value in the
+            /// constructor, and set to <see langword="null"/> when the current reference is disposed.
+            /// </summary>
+            /// <remarks>
+            /// <para>This value is only cleared in order to support cases where one or more references is garbage
+            /// collected without having <see cref="Dispose"/> called.</para>
+            /// </remarks>
+            private T _instance;
+
+            /// <summary>
+            /// The boxed reference count, which is shared by all references with the same <see cref="Target"/> object.
+            /// </summary>
+            /// <remarks>
+            /// <para>This field serves as the synchronization object for the current type, since it is shared among all
+            /// counted reference to the same target object. Accesses to <see cref="StrongBox{T}.Value"/> should only
+            /// occur when this object is locked.</para>
+            ///
+            /// <para>PERF DEV NOTE: A concurrent (but complex) implementation of this type with identical semantics is
+            /// available in source control history. The use of exclusive locks was not causing any measurable
+            /// performance overhead even on 28-thread machines at the time this was written.</para>
+            /// </remarks>
+            private readonly StrongBox<int> _boxedReferenceCount;
+
+            /// <summary>
+            /// Initializes a new reference counting wrapper around an <see cref="IDisposable"/> object.
+            /// </summary>
+            /// <remarks>
+            /// <para>The reference count is initialized to 1.</para>
+            /// </remarks>
+            /// <param name="instance">The object owned by this wrapper.</param>
+            public ReferenceCountedDisposable(T instance)
+            {
+                _instance = instance ?? throw new ArgumentNullException(nameof(instance));
+                _boxedReferenceCount = new StrongBox<int>(1);
+            }
+
+            private ReferenceCountedDisposable(T instance, StrongBox<int> referenceCount)
+            {
+                _instance = instance;
+
+                // The reference count has already been incremented for this instance
+                _boxedReferenceCount = referenceCount;
+            }
+
+            /// <summary>
+            /// Gets the target object.
+            /// </summary>
+            /// <remarks>
+            /// <para>This call is not valid after <see cref="Dispose"/> is called. If this property or the target
+            /// object is used concurrently with a call to <see cref="Dispose"/>, it is possible for the code to be
+            /// using a disposed object. After the current instance is disposed, this property returns
+            /// <see langword="null"/>. However, the exact time when this property starts returning null after
```

Comment:
Maybe another thought: you can't be safe when people misuse you, but it's probably better to let them not misuse you then. :smile: